lib/posix/pthread.c

```diff
@@ -302,6 +302,21 @@ int pthread_getschedparam(pthread_t pthread, int *policy,
 	return 0;
 }
 
+/**
+ * @brief Dynamic package initialization
+ *
+ * See IEEE 1003.1
+ */
+int pthread_once(pthread_once_t *once, void (*initFunc)(void))
+{
+	if (*once == PTHREAD_ONCE_INIT) {
```

Comment:
This is racy: please use `atomic.h` here.  What you want is a `compare-and-set` primitive (`atomic_cas()` in `atomic.h`):  if you successfully compare `*once` with, say, a "not initialized" value, swap it with a "initialized value" and call the init function.  It is close to what you're doing here, but this operation must be atomic; otherwise, `initFunc()` may be called twice if this code is preempted midway.

Reply:
This might work:

```C
#define NEVER 0
#define INITIALIZING 1
#define INITIALIZED 2

void once(atomic_val_t *once, void (*fn)(void)) {
  /* Nothing to do: fn() has been called once according to the *once
   * value. */
  if (atomic_get(once) == INITIALIZED) {
      return;
  }

  /* Try going from "never ran" to "will run"... */
  if (atomic_cas(once, NEVER, INITIALIZING)) {
      fn();
      /* ...and then go to "we executed" once fn() has executed */
      atomic_inc(once);
      return;
  }

  /* atomic_cas() returned false; in this case, it's in either
   * INITIALIZING or INITIALIZED at this point.  Spin on it
   * until it changes its value to INITIALIZED.
   */
  while (atomic_get(once) != INITIALIZED) {
      k_yield();
  }
}
```

Reply:
@andyross Could you take a look here?  Do you see anything wrong with this `once()` implementation above?

Reply:
That would work, but spinning on yield is just a bad idea.  The implication from way it's documented is that any blocking will be OS-managed.   You can do this with a condition variable relatively easily, just implement this in terms of other POSIX APIs..  Put a pthread_mutex around the "once" accesses, and in the loop at the end do ptherad_cond_wait() instead of yield, with a pthread_cond_broadcast() after fn() has returned.

Reply:
@lpereira Thanks a lot for explaining this in good detail, I can see the importance of handling race condition now. Shall I go ahead and use POSIX APIs to implement this, like Andy has mentioned?